### ISSUE TITLE
TCP Keep-Alive

### ISSUE THREAD

**Issue by zmhh** [Azure/amqpnetlite#234]
I am still working on getting a code sample that demonstrates the issue, but long story short I have observed issues using this library with bad network connections wherein sockets are left alive in the "ESTABLISHED" state, even though they are no longer used.

From the [AMQP specification](http://docs.oasis-open.org/amqp/core/v1.0/cos01/amqp-core-transport-v1.0-cos01.html#doc-doc-idle-time-out):

"The use of idle timeouts is in addition to any network protocol level control. Implementations SHOULD make use of TCP keep-alive wherever possible in order to be good citizens."

From what I can see, this library does not yet make use of TCP keep-alive.  I believe this should be implemented, and the keep-alive period controlled by the developer.

**Comment by zmhh** [comment]
The following code, combined with usage of the network simulation tool [Clumsy](https://jagt.github.io/clumsy/), demonstrates the issue.

Client:
`class Program
{

    static void Main(string[] args)
    {
        var clients = new Task[1];

        for (int i = 0; i < clients.Length; i++)
        {
            clients[i] = Task.Run(async () => await WriteForeverAsync());
        }

        Task.WaitAny(clients);
    }

    private const int Timeout = 5000;

    static async Task WriteForeverAsync()
    {
        var id = Guid.NewGuid().ToString();
        var msg = 0;

        Connection connection = null;
        Session session = null;
        SenderLink senderlink = null;

        try
        {
            while (true)
            {
                await Task.Delay(100);

                try
                {
                    if (senderlink == null)
                    {
                        ConnectionFactory connFactory = new ConnectionFactory();
                        connFactory.AMQP.MaxFrameSize = 5242880;

                        connection = connFactory.CreateAsync(new Address("amqp://destinationMachine:8881")).Result;
                        session = new Session(connection);
                        senderlink = new SenderLink(session, id, "ingress_queue");
                    }

                    Console.WriteLine($"{id} : sending...");
                    await Task.Run(() => senderlink.Send(new Message(msg.ToString()), Timeout));
                    msg++;
                }
                catch (AmqpException e)
                {
                    Console.WriteLine($"{id} : {e.Error.Condition}, {e.Error.Description}");

                    senderlink?.Close(Timeout);
                    session?.Close(Timeout);
                    connection?.Close(Timeout);

                    senderlink = null;
                }
                catch (Exception e)
                {
                    Console.WriteLine($"{id} : {e}");

                    senderlink?.Close(Timeout);
                    session?.Close(Timeout);
                    connection?.Close(Timeout);

                    senderlink = null;
                }


                Console.WriteLine($"{id} : end");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine($"{id} : UNEXPECTED, {e}");
        }
    }
}`

Server:
`class Program
{

    static void Main(string[] args)
    {
        var clients = new Task[1];

        for (int i = 0; i < clients.Length; i++)
        {
            clients[i] = Task.Run(async () => await WriteForeverAsync());
        }

        Task.WaitAny(clients);
    }

    private const int Timeout = 5000;

    static async Task WriteForeverAsync()
    {
        var id = Guid.NewGuid().ToString();
        var msg = 0;

        Connection connection = null;
        Session session = null;
        SenderLink senderlink = null;

        try
        {
            while (true)
            {
                await Task.Delay(100);

                try
                {
                    if (senderlink == null)
                    {
                        ConnectionFactory connFactory = new ConnectionFactory();
                        connFactory.AMQP.MaxFrameSize = 5242880;

                        connection = connFactory.CreateAsync(new Address("amqp://mhh-dev:8881")).Result;
                        session = new Session(connection);
                        senderlink = new SenderLink(session, id, "ingress_queue");
                    }

                    Console.WriteLine($"{id} : sending...");
                    await Task.Run(() => senderlink.Send(new Message(msg.ToString()), Timeout));
                    msg++;
                }
                catch (AmqpException e)
                {
                    Console.WriteLine($"{id} : {e.Error.Condition}, {e.Error.Description}");

                    senderlink?.Close(Timeout);
                    session?.Close(Timeout);
                    connection?.Close(Timeout);

                    senderlink = null;
                }
                catch (Exception e)
                {
                    Console.WriteLine($"{id} : {e}");

                    senderlink?.Close(Timeout);
                    session?.Close(Timeout);
                    connection?.Close(Timeout);

                    senderlink = null;
                }


                Console.WriteLine($"{id} : end");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine($"{id} : UNEXPECTED, {e}");
        }
    }`


To see the issue, run the client and server programs on different machines.  Allow the connection to be established and for data to begin sending.

Observe the single socket open on the server machine.  On Windows you can do this with "netstat -an".

Then using Clumsy on the Server machine, drop all inbound and outbound packets.  Leave this for a minute or so.  Careful not to lock yourself out of the machine if it's remote.  I used the following settings:

![image](https://user-images.githubusercontent.com/20959889/33093882-19693ae6-ceb3-11e7-87cf-6387f6128031.png)

Finally disable Clumsy, restoring the network connection, and allow the client to reconnect.  Observe there are now two established sockets, only one being used.  The old socket is left in the "ESTABLISHED" state until the application terminates.

![image](https://user-images.githubusercontent.com/20959889/33093855-011a4af2-ceb3-11e7-8343-e884654222c3.png)

One reason this can be a severe issue, is that if the server machine is listening to multiple clients with frequent dropped connections, the server machine could eventually run out of sockets.

A TCP KeepAlive would prevent this issue, as well as other potential issues when a network connection misbehaves.

**Comment by zmhh** [comment]
I have created a pull request for a change which addresses this issue: TCP Keepalive added as TcpSetting #235

https://github.com/Azure/amqpnetlite/pull/235

**Comment by xinchen10** [comment]
@zmhh the Server code is the same as the client? Was that a copy/paste error? The library supports amqp level connection idle timeout. The broken connection can be detected when the idle-timeout is set even without tcp keepalive option. But I can see the value of supporting tcp keep-alive so user has another choice. The following code shows how to set amqp connection idle-timeout on the listener. If a heartbeat is not received within the specified time, the server will close the socket.

```
this.host = new ContainerHost(new Uri("amqp://localhost:5672"));
this.host.Listeners[0].AMQP.IdleTimeout = 60000;
this.host.Open();
```

**Comment by zmhh** [comment]
@xinchen10 Yes that was a copy paste error, sorry.  I've updated the code sample in the original post to have the actual server code.  I also updated the client code to work with the current master version of amqpnetlite (previously it was using 1.2.2)

The IdleTimeout parameter appears to work fine for the client, but does not have any effect on the listener.  I've included an IdleTimeout in the Server code, which does not prevent the issue.  This may be a separate bug.
